### PR TITLE
Validate data dir in database copy

### PR DESCRIPTION
**Pull request by jkosh44**
Testdrive has an option, --validate-data-dir, that validates the
contents of the SQLite database. It does this by opening a new catalog
using the database directory and comparing the contents against the in
memory catalog.

Opening a new catalog using the same SQLite database that Materialize
is actively using can cause some conflicting database operations.
Therefore this commit updates Testdrive to open a new catalog using a
copy of the SQLite database to avoid conflicts.

### Motivation
This PR adds a feature that has not yet been specified.

### Testing

- [X] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/testdrive/src/action/sql.rs [hunk 1]
```diff
@@ -190,6 +191,11 @@ impl Action for SqlAction {
                 | Statement::CreateView { .. }
                 | Statement::DropDatabase { .. }
                 | Statement::DropObjects { .. } => {
+                    let temp_dir = tempfile::tempdir()?;
+                    let mut options = CopyOptions::new();
+                    options.content_only = true;
+                    fs_extra::dir::copy(&path, temp_dir.path(), &options)?;
+                    let path = temp_dir.path();
```
Comment (benesch) [inline src/testdrive/src/action/sql.rs:198]: Can you add a comment about why we do this?

File: src/testdrive/src/action/verify_timestamp_compaction.rs [hunk 2]
```diff
@@ -56,6 +57,11 @@ impl Action for VerifyTimestampCompactionAction {
 
     async fn redo(&self, state: &mut State) -> Result<ControlFlow, anyhow::Error> {
         if let Some(path) = &state.materialized_data_path {
+            let temp_dir = tempfile::tempdir()?;
+            let mut options = CopyOptions::new();
+            options.content_only = true;
+            fs_extra::dir::copy(&path, temp_dir.path(), &options)?;
+            let path = temp_dir.path();
```
Comment (benesch) [inline src/testdrive/src/action/verify_timestamp_compaction.rs:64]: Ditto.

**Review by benesch — APPROVED** [review]
LGTM if it looks good to CI.

File: src/testdrive/src/action/sql.rs (hunk 1 above)
Comment (jkosh44) [reply src/testdrive/src/action/sql.rs]: I extracted this out to it's own function/mod and put a comment there.

**Comment by jkosh44** [comment]
@benesch I cannot figure this out. The issue is that the `/share/mzdata` volume doesn't have the right permissions so the copy from mzdata to the temp directory fails. 

I can fix this locally by manually changing the docker volumes permission on my computer (`sudo chown -R $USER:docker /var/lib/docker/volumes/testdrive_mzdata`) but I'm a little stumped as to how to do this for CI.

**Review by benesch — APPROVED** [review]
This time for sure! 🤞🏽

**Pull request merged**